apps/aetx/src/aetx_sign.erl

```diff
@@ -104,71 +105,85 @@ from_db_format(#signed_tx{tx = Tx} = STx) ->
+    Bin = aetx:serialize_to_binary(Tx),
     case aetx:signers(Tx, Trees) of
         {ok, Signers} ->
-            verify_signatures(Signers, Bin, Sigs);
+            verify_signatures(Signers, Bin, Sigs, Height);
         {error, _Reason} ->
             {error, signature_check_failed}
     end.
 
 %% this function is strict and does not allow having more signatures that the
 %% one being checked
--spec verify_half_signed(aec_keys:pubkey() | [aec_keys:pubkey()],
-                         signed_tx()) -> ok | {error, signature_check_failed}.
-verify_half_signed(Signer, SignedTx) when is_binary(Signer) ->
-    verify_half_signed([Signer], SignedTx);
-verify_half_signed(Signers, #signed_tx{tx = Tx, signatures = Sigs}) ->
-    Bin     = aetx:serialize_to_binary(Tx),
-    verify_signatures(Signers, Bin, Sigs).
 
 %% this function allows having more signatures that the one being checked
--spec verify_one_pubkey(aec_keys:pubkey(), signed_tx())
-    -> ok | {error, signature_check_failed}.
-verify_one_pubkey(Signer, #signed_tx{tx = Tx, signatures = Sigs}) ->
+-spec verify_one_pubkey(aec_keys:pubkey(), signed_tx(), aec_blocks:height()) ->
+    ok | {error, signature_check_failed}.
+verify_one_pubkey(Signer, #signed_tx{tx = Tx, signatures = Sigs}, Height) ->
     Bin = aetx:serialize_to_binary(Tx),
-    case verify_one_pubkey(Sigs, Signer, Bin) of
+    case verify_one_pubkey(Sigs, Signer, Bin, Height) of
         {ok, _} -> ok;
         error -> {error, signature_check_failed}
     end.
 
-verify_signatures([PubKey|Left], Bin, Sigs) ->
-    case verify_one_pubkey(Sigs, PubKey, Bin) of
-        {ok, SigsLeft} -> verify_signatures(Left, Bin, SigsLeft);
-        error -> {error, signature_check_failed}
-    end;
-verify_signatures([],_Bin, []) ->
+-spec verify_half_signed(aec_keys:pubkey() | [aec_keys:pubkey()],
+                         signed_tx(), aec_blocks:height()) ->
+    ok | {error, signature_check_failed}.
+verify_half_signed(Signer, SignedTx, Height) when is_binary(Signer) ->
+    verify_half_signed([Signer], SignedTx, Height);
+verify_half_signed(Signers, #signed_tx{tx = Tx, signatures = Sigs}, Height) ->
+    verify_signatures(Signers, aetx:serialize_to_binary(Tx), Sigs, Height).
+
+verify_signatures([], _Bin, [], _Height) ->
     ok;
-verify_signatures(PubKeys,_Bin, Sigs) ->
+verify_signatures([PubKey|Left], Bin, Sigs, Height) ->
+    case verify_one_pubkey(Sigs, PubKey, Bin, Height) of
+        {ok, SigsLeft} -> verify_signatures(Left, Bin, SigsLeft, Height);
+        error          -> {error, signature_check_failed}
+    end;
+verify_signatures(PubKeys,_Bin, Sigs, _Height) ->
     lager:debug("Signature check failed: ~p ~p", [PubKeys, Sigs]),
     {error, signature_check_failed}.
 
-verify_one_pubkey(Sigs, PubKey, Bin) ->
-    verify_one_pubkey(Sigs, PubKey, Bin, []).
+verify_one_pubkey(Sigs, PubKey, Bin, Height) when ?VALID_PUBK(PubKey) ->
+    Protocol = aec_hard_forks:protocol_effective_at_height(Height),
```

Comment:
I would have preferred awareness to protocol - rather than height so deep in call - on the basis that support for miner signalling consensus upgrade may remove assumption that consensus is known based on height. Though ok.

Reply:
This was done a bit before we started discussing the signalling, sorry...